### PR TITLE
Concat check input size check

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Each module output inside concat is supposed to have the same dim size exception the one for concatenating, this change is to do the sanity check to make it more user friendly when exception happens

## How was this patch tested?

Unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/2027

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala [hunk 1]
```diff
@@ -108,6 +110,27 @@ class Concat[T: ClassTag](val dimension: Int)(
     this.output
   }
 
+  private def checkOutputs(outputs : Array[Tensor[T]], dim : Int) : Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: Can we do this check at line 66? when update this.size also check other dimensinos

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala [hunk 2]
```diff
@@ -108,6 +110,27 @@ class Concat[T: ClassTag](val dimension: Int)(
...
+    if (outputs == null || outputs.length == 0) {
+      return
+    }
+    val size = outputs(0).size()
+    val dim = size.length
+    for ( i <- 1 until outputs.length) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: avoid to use for here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala (hunk 1 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: @yiheng  makes sense,  we don't need to wait all module to compute before check, coming back shortly with refinement

**Comment by wzhongyuan** [comment]
@yiheng  pls review the updates

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala [hunk 3]
```diff
@@ -64,6 +64,16 @@ class Concat[T: ClassTag](val dimension: Int)(
         this.size = currentOutput.size()
       } else {
         this.size(this.dimension - 1) += currentOutput.size(this.dimension)
+        require(this.size.length == currentOutput.size.length,
+        s"${this.modules(i).getName} output size mismatch, expected : ${this.size.length}," +
+          s"actual ${currentOutput.size.length}")
+        this.size.zipWithIndex.foreach(size => {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Concat.scala]: try to use while instead of foreach for performance concern...

**Comment by yiheng** [comment]
also add a unit test to verfy the exception?

**Comment by wzhongyuan** [comment]
@yiheng  Refinement on top of your comments, please review

**Comment by yiheng** [comment]
LGTM

**Comment by wzhongyuan** [comment]
PR validation passed http://172.168.2.101:8080/job/BigDL-PR-V/560/

**Pull request merged**